clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1461,11 +1462,28 @@ private static String partitionLeadMetricName(TopicPartition tp) {
 
     }
 
+    private void closeFetchSessions() {
+        Cluster cluster = metadata.fetch();
+
+        for (Map.Entry<Integer, FetchSessionHandler> entry : sessionHandlers.entrySet()) {
+            Integer nodeId = entry.getKey();
+            int sessionId = entry.getValue().sessionId();
+            FetchMetadata closeSessionMetadata = new FetchMetadata(sessionId, FetchMetadata.FINAL_EPOCH);
+
+            final FetchRequest.Builder closeSessionRequest = FetchRequest.Builder
+                    .forConsumer(this.maxWaitMs, this.minBytes, Collections.emptyMap())
+                    .metadata(closeSessionMetadata);
+            client.send(cluster.nodeById(nodeId), closeSessionRequest);
```

Comment:
Don't we have to do something with the futures returned by `send`? How do we know if the requests completed? Also, cc @hachikuji.

Reply:
We had discussed offline that there is no need to wait and ensure the requests completed. It makes sense to make it best effort, otherwise it can slow down client termination and I assume people prefer the client terminates quicker rather than make sure no unused sessions are in the broker

Reply:
We should always document things like that so that people reading the code are aware of it. My point was more subtle though, how do we make sure we actually sent the request?

Reply:
More specifically, there's no point in having this if we close the network client before we even get to the point where the request has reached the network layer.

Reply:
Oh, I did not know that was a possibility. I'll dive in.
Fair point with the documentation

Reply:
It probably makes sense to wait a short time, but not very long, to send the request to the broker.  I'm not sure how that interacts with the rest of the client close logic

Reply:
Yeah, good catch @ijuma. This is the code for closing the `KafkaConsumer`:
```java
ClientUtils.closeQuietly(fetcher, "fetcher", firstException);
ClientUtils.closeQuietly(interceptors, "consumer interceptors", firstException);
ClientUtils.closeQuietly(metrics, "consumer metrics", firstException);
ClientUtils.closeQuietly(client, "consumer network client", firstException);
ClientUtils.closeQuietly(keyDeserializer, "consumer key deserializer", firstException);
ClientUtils.closeQuietly(valueDeserializer, "consumer value deserializer", firstException);
```
Meaning there's a pretty big chance the client is closed before the request are sent.

But then I read the Javadoc of `#send()`
```
Send a new request. Note that the request is not actually transmitted on the
     * network until one of the {@link #poll(long)} variants is invoked. At this
     * point the request will either be transmitted successfully or will fail.
     * Use the returned future to obtain the result of the send. Note that there is no
     * need to check for disconnects explicitly on the {@link ClientResponse} object;
     * instead, the future will be failed with a {@link DisconnectException}.
```
Meaning this code doesn't transmit any requests since `poll` is not called at all...

I see two approaches to this
1) Call `poll(Long timeout)` with the `request.timeout.ms` configured timeout after the for loop. This will give all requests a grand total of `timeout` time to be sent. This call will block for a total maximum of `request.timeout.ms` (default is 30 seconds which seems high to me for this use case). Maybe a total timeout of N seconds, where N is the number of requests?
2) Call `poll((RequestFuture<?> future, long timeout)` individually for each sent close request. This blocks on each call, so it's questionable what the timeout should be (1 second?)

In both cases, we will slow down the closing of the consumer for a bit, as we will have to block as far as I can tell.

As a sidenote - this code would also need to handle exceptions as it currently doesn't. My idea is to only catch & log the error.

Reply:
The consumer close code needs to block until the fetch sessions are closed.  `KafkaConsumer` has `close()` and `close(long timeout, TimeUnit timeUnit)`.  In the case of the latter one, we don't want to wait longer than the specified time.